File: web-ts/src/state/activeHubConnection.ts
// CYP-853 (Multi-Hub M3) — the switch-first / ONE-ACTIVE-HUB connection controller (CYP-748-Q4 ratified). The active
// pointer drives EXACTLY ONE live connection: one createRemoteConnMachine (CYP-822/826) + one statusFeed (CYP-844)
// behind an injected connector seam. Switching TEARS DOWN the previous hub's connection before opening the new one —
// inactive hubs hold NO live background connection (resource-safe, CYP-611 sizing; and the precondition for the
// immediate STALE lapse in hubTrustProvenance.ts — no connection ⇒ no observation).
//
// The connector is INJECTED: in tests it's a fake that hands back a machine driven by FakeRemoteConnector; the real
// dial (socket + statusFeed over the wire) is the ARMING seam (CYP-807-A5), not wired here. This controller owns only
// the ONE-ACTIVE lifecycle, not the transport.
import type { RemoteConnMachine } from './remoteConnState'
import type { HubId } from '../net/hubRegistry'

/** A single hub's live connection. `close()` tears it down (socket/feed closed, no further reconnect). */
export interface HubConnectionHandle {
  readonly machine: RemoteConnMachine
  close(): void
}

/** How to OPEN a live connection for a hub. Real = dial + statusFeed (arming); test/stub = FakeRemoteConnector-driven. */
export type HubConnector = (hubId: HubId) => HubConnectionHandle

export interface ActiveHubConnection {
  /** The currently active hub + its machine, or null before the first connect. */
  active(): { readonly hubId: HubId; readonly machine: RemoteConnMachine } | null
  /**
   * Make `hubId` the active hub: tear down the previous hub's live connection FIRST (no background connection survives
   * to an inactive hub), then open the new one. Switching to the already-active hub is a no-op (the live connection is
   * NOT torn down and re-dialed — mirrors the switcher's switch-to-active no-op).
   */
  switchTo(hubId: HubId): void
  /** Tear down the active connection entirely (App unmount). */
  close(): void
}

export function createActiveHubConnection(connector: HubConnector): ActiveHubConnection {
  let current: { hubId: HubId; handle: HubConnectionHandle } | null = null

  return {
    active: () => (current === null ? null : { hubId: current.hubId, machine: current.handle.machine }),
    switchTo: (hubId) => {
      if (current !== null && current.hubId === hubId) return // switch-to-active = no-op (do NOT tear down + re-dial)
      current?.handle.close() // tear down the OLD first — exactly one live connection at a time
      current = { hubId, handle: connector(hubId) }
    },
    close: () => {
      current?.handle.close()
      current = null
    },
  }
}

File: app/shared/src/commonMain/kotlin/com/tneff/cyppieagents/multihub/ActiveHubConnection.kt
package com.tneff.cyppieagents.multihub

/**
 * CYP-865 (Compose-M3-Mirror of web-ts CYP-853) — the **switch-first / ONE-ACTIVE-HUB** connection controller
 * (CYP-748-Q4). The active pointer drives EXACTLY ONE live connection behind an **injected** [HubConnector];
 * switching TEARS DOWN the previous hub's connection BEFORE opening the new one, so inactive hubs hold **no** live
 * background connection (resource-safe, CYP-611 sizing; and the precondition for the immediate STALE lapse in
 * [displayedTrust] — no connection ⇒ no observation).
 *
 * **Arming DARK (§9.3).** The [HubConnector] is INJECTED: tests/stubs hand back a fake handle; the **real** dial
 * (socket + statusFeed over the wire) is the ARMING seam (CYP-807-A5), NOT wired here — this controller owns only
 * the ONE-ACTIVE lifecycle, not the transport. The connection's machine type is left generic ([M]): Compose has no
 * `RemoteConnState` machine yet (a separate CYP-822/826-class unit) — the concrete state arrives with M5/arming.
 */

/** A single hub's live connection. [close] tears it down (socket/feed closed, no further reconnect). */
interface HubConnectionHandle<out M> {
    val machine: M
    fun close()
}

/** How to OPEN a live connection for a hub. Real = dial + statusFeed (arming); test/stub = fake-driven. INJECTED. */
fun interface HubConnector<M> {
    fun open(hubId: String): HubConnectionHandle<M>
}

/** The currently-active hub + its machine. */
data class ActiveHub<out M>(val hubId: String, val machine: M)

interface ActiveHubConnection<M> {
    /** The currently active hub + its machine, or null before the first connect. */
    fun active(): ActiveHub<M>?

    /**
     * Make [hubId] the active hub: tear down the previous hub's live connection FIRST (no background connection
     * survives to an inactive hub), THEN open the new one. Switching to the ALREADY-active hub is a **no-op** — the
     * live connection is NOT torn down and re-dialed (mirrors the switcher's switch-to-active no-op).
     */
    fun switchTo(hubId: String)

    /** Tear down the active connection entirely (host unmount). */
    fun close()
}

fun <M> createActiveHubConnection(connector: HubConnector<M>): ActiveHubConnection<M> =
    object : ActiveHubConnection<M> {
        private var current: Pair<String, HubConnectionHandle<M>>? = null

        override fun active(): ActiveHub<M>? =
            current?.let { (hubId, handle) -> ActiveHub(hubId, handle.machine) }

        override fun switchTo(hubId: String) {
            if (current?.first == hubId) return // switch-to-active = no-op (do NOT tear down + re-dial)
            current?.second?.close() // tear down the OLD first — exactly one live connection at a time
            current = hubId to connector.open(hubId)
        }

        override fun close() {
            current?.second?.close()
            current = null
        }
    }
